logger_bench_test.go

```diff
@@ -238,3 +240,130 @@ func Benchmark100Fields(b *testing.B) {
 		logger.With(first...).Info("Child loggers with lots of context.", second...)
 	}
 }
+
+func dummy(wg *sync.WaitGroup, s string, i int) string {
+	if i == 0 {
+		wg.Wait()
+		return "1" + s
+	}
+	return dummy(wg, s, i-1)
+}
+
+func stackGrower(n int) *sync.WaitGroup {
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	go dummy(&wg, "hi", n)
+	return &wg
+}
+
+func BenchmarkAny(b *testing.B) {
+	var (
+		logger = New(
+			zapcore.NewCore(
+				zapcore.NewJSONEncoder(NewProductionConfig().EncoderConfig),
+				&ztest.Discarder{},
+				DebugLevel,
+			),
+		)
+		key = "some-long-string-longer-than-16"
+	)
+
+	tests := []struct {
+		name   string
+		typed  func() Field
+		anyArg any
+	}{
+		{
+			name:   "string",
+			typed:  func() Field { return String(key, "yet-another-long-string") },
+			anyArg: "yet-another-long-string",
+		},
+		{
+			name:   "stringer",
+			typed:  func() Field { return Stringer(key, InfoLevel) },
+			anyArg: InfoLevel,
+		},
+	}
+
+	for _, tt := range tests {
+		b.Run(tt.name+"-typ-no-logger", func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				f := tt.typed()
+				runtime.KeepAlive(f)
+			}
+		})
+		b.Run(tt.name+"-any-no-logger", func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				f := Any(key, tt.anyArg)
+				runtime.KeepAlive(f)
+			}
+		})
+		b.Run(tt.name+"-typ-logger", func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				logger.Info("", tt.typed())
+			}
+		})
+		b.Run(tt.name+"-any-logger", func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				logger.Info("", Any(key, tt.anyArg))
+			}
```

Comment:
For these two cases, we might want to use withBenchedLogger.

https://github.com/abhinav/zap/blob/1058d37e5320ddd1a6da01e76247fa64dd8ad26d/logger_bench_test.go#L53

That's used in existing logging benchmarks:

https://github.com/abhinav/zap/blob/1058d37e5320ddd1a6da01e76247fa64dd8ad26d/logger_bench_test.go#L155-L159

Although it'll use `b.RunParallel` -- I don't know if that's okay?

Reply:
Yeah, I don't know if it matters.

Looking at https://pkg.go.dev/testing#B.RunParallel, it means we'll start a GOMAXPROCS goroutines, and then distribute the actual tests (which will start their own goroutines) there.

Does it matter? Does it not matter? I can test/try.

Reply:
Oh, just meant these two cases, rather than all loggers.

```
BenchmarkAny/string-typ-logger                   2959306               406.9 ns/op            64 B/op          1 allocs/op
BenchmarkAny/string-typ-logger-12               15433677                80.36 ns/op           64 B/op          1 allocs/op
BenchmarkAny/string-any-logger                   2845083               413.0 ns/op            64 B/op          1 allocs/op
BenchmarkAny/string-any-logger-12               15772636                79.69 ns/op           64 B/op          1 allocs/op
```

the difference for 12- gomacprocs is huge.

I guess it makes sense to include it like this.

Reply:
Added to all tests.